modules/imgproc/src/warp_kernels.simd.hpp

```diff
@@ -2130,6 +2195,2067 @@ void warpAffineLinearApproxInvoker_8UC4(const uint8_t *src_data, size_t src_step
+        }
+        v_uint8 bval_v0 = vx_load_low(&bvalbuf[0]);
+        v_uint8 bval_v1 = vx_load_low(&bvalbuf[uf]);
+        v_uint8 bval_v2 = vx_load_low(&bvalbuf[uf*2]);
+        uint8x8_t reds = {0, 8, 16, 24, 3, 11, 19, 27},
+                  greens = {1, 9, 17, 25, 4, 12, 20, 28},
+                  blues = {2, 10, 18, 26, 5, 13, 21, 29};
+
+        for (int y = r.start; y < r.end; y++) {
+            uint8_t* dstptr = dst + y*dststep;
+            int x = 0;
+
+            v_float32 dst_x0 = vx_load(start_indices.data());
+            v_float32 dst_x1 = v_add(dst_x0, vx_setall_f32(float(vlanes_32)));
+            v_float32 M0 = vx_setall_f32(M[0]),
+                      M3 = vx_setall_f32(M[3]),
+                      M6 = vx_setall_f32(M[6]);
+            v_float32 M_x = vx_setall_f32(static_cast<float>(y * M[1] + M[2])),
+                      M_y = vx_setall_f32(static_cast<float>(y * M[4] + M[5])),
+                      M_w = vx_setall_f32(static_cast<float>(y * M[7] + M[8]));
+
+            for (; x < dstcols - uf; x += uf) {
+                // [TODO] apply halide trick
+
+                CV_WARPPERSPECTIVE_LINEAR_VECTOR_COMPUTE_MAPPED_COORD();
+
+                v_int32 addr_0 = v_fma(v_srcstep, src_iy0, v_mul(src_ix0, three)),
+                        addr_1 = v_fma(v_srcstep, src_iy1, v_mul(src_ix1, three));
+                vx_store(addr, addr_0);
+                vx_store(addr + vlanes_32, addr_1);
+
+                uint8x8_t p00r, p01r, p10r, p11r,
+                          p00g, p01g, p10g, p11g,
+                          p00b, p01b, p10b, p11b;
+
+                if (v_reduce_min(inner_mask) != 0) { // all loaded pixels are completely inside the image
+                    uint8x8x4_t p00 = {
+                        vld1_u8(src + addr[0]),
+                        vld1_u8(src + addr[1]),
+                        vld1_u8(src + addr[2]),
+                        vld1_u8(src + addr[3])
+                    };
+
+                    uint8x8x4_t p01 = {
+                        vld1_u8(src + addr[4]),
+                        vld1_u8(src + addr[5]),
+                        vld1_u8(src + addr[6]),
+                        vld1_u8(src + addr[7])
+                    };
+
+                    uint8x8x4_t p10 = {
+                        vld1_u8(src + addr[0] + srcstep),
+                        vld1_u8(src + addr[1] + srcstep),
+                        vld1_u8(src + addr[2] + srcstep),
+                        vld1_u8(src + addr[3] + srcstep)
+                    };
+
+                    uint8x8x4_t p11 = {
+                        vld1_u8(src + addr[4] + srcstep),
+                        vld1_u8(src + addr[5] + srcstep),
+                        vld1_u8(src + addr[6] + srcstep),
+                        vld1_u8(src + addr[7] + srcstep)
+                    };
+
+                    uint32x2_t p00_, p01_, p10_, p11_;
+
+                    p00_ = vreinterpret_u32_u8(vtbl4_u8(p00, reds));
```

Comment:
lot's of, lot's of duplicates. Let's share all such kernels between warpAffine, warpPerspective and the future remap.

Reply:
Yup, done with macros.